### PR TITLE
Add NodeOutboundIPs customization in Azure cluster

### LINKED ISSUE #1394: Support existing IP as node outbound IPs
/kind feature

**Describe the solution you'd like**

In many cases user would like to configure the cluster outbound IP to a set of pre-created static IPs or IP prefixes. External systems like SQL firewall will whitelist this IP in its ACL. 

We should allow customization of node outbound IPs to reuse existing IPs. This way new clusters, or re-created clusters can reuse the same set of IP, such that all remote ACLs won't need update.

Interface would be a new `outboundIPs` property in the `NodeOutboundLB` property from azurecluster

Summary
* Use outbound IP as ACL is common practice
* Modifying those ACLs are difficult, sometimes impossible
* Re-create cluster / migrating old clusters (from self-managed) to cluster API is common scenario
* To allow reusing same ACL, we should allow reuse existing IP as outbound

**Anything else you would like to add:**

**Environment:**

- cluster-api-provider-azure version: any
- Kubernetes version: (use `kubectl version`): any
- OS (e.g. from `/etc/os-release`): Ubuntu

### PR DESCRIPTION
**Pull request by Ankitasw**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR allows to add `frontendIPs` to `NodeOutboundLB` to allow users to use existing IPs as nodeoutbound IPs

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1394 

**Special notes for your reviewer**:

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:
```release-note
Add support for using existing IPs as nodeoutbound IPs in Azure cluster
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
/hold cancel

#1740 has merged, please rebase and move api changes to v1beta1

**Comment by CecileRobertMichon** [comment]
> Adding outboundIP property to FrontendIP.PublicIP seems fair to me. @CecileRobertMichon do you have any thoughts?

Sorry I missed this earlier. +1 to Shyam's suggestion.

File: api/v1beta1/azurecluster_validation.go [hunk 1]
```diff
@@ -498,3 +508,12 @@ func validateCloudProviderConfigOverrides(old, new *CloudProviderConfigOverrides
 	}
 	return allErrs
 }
+
+func validateOutboundIP(frontendIP FrontendIP) error {
+	outboundIP := net.ParseIP(frontendIP.PublicIP.OutboundIP)
+	_, _, err := net.ParseCIDR(frontendIP.PublicIP.OutboundIP)
+	if outboundIP == nil && err != nil {
+		return err
```
Comment (shysank) [inline api/v1beta1/azurecluster_validation.go:516]: can we wrap this into a error that says something about `outboundIP==nil` as well?

File: api/v1beta1/azurecluster_validation_test.go [hunk 2]
```diff
@@ -1024,9 +1024,11 @@ func TestValidateNodeOutboundLB(t *testing.T) {
 			name: "FrontendIps can update when frontendIpsCount changes",
 			lb: &LoadBalancerSpec{
 				FrontendIPs: []FrontendIP{{
-					Name: "some-frontend-ip-1",
+					Name:             "some-frontend-ip-1",
+					PrivateIPAddress: "1.2.3.4",
```
Comment (shysank) [inline api/v1beta1/azurecluster_validation_test.go:1028]: is this test related to the changes?

File: api/v1beta1/azurecluster_validation_test.go [hunk 3]
```diff
@@ -1024,9 +1024,11 @@ func TestValidateNodeOutboundLB(t *testing.T) {
...
 				}, {
-					Name: "some-frontend-ip-2",
+					Name:             "some-frontend-ip-2",
+					PrivateIPAddress: "5.6.7.8",
```
Comment (shysank) [inline api/v1beta1/azurecluster_validation_test.go:1031]: same comment as above

File: api/v1beta1/azurecluster_validation.go [hunk 4]
```diff
@@ -498,3 +508,12 @@ func validateCloudProviderConfigOverrides(old, new *CloudProviderConfigOverrides
...
+	}
+	return nil
```
Comment (shysank) [inline api/v1beta1/azurecluster_validation.go:518]: wdyt about this?
```suggestion
	outboundIP := net.ParseIP(frontendIP.PublicIP.OutboundIP)
	if outboundIP != nil {
	   return nil
	}
	
	_, _, err := net.ParseCIDR(frontendIP.PublicIP.OutboundIP)
	if err == nil {
	   return nil
	}
	
	return errors.Wrap("could not parse into an ip or prefix", err)
```

Comment (shysank) [reply api/v1beta1/azurecluster_validation.go:518]: or atleast wrap the error with information about `outboundIP == nil` condition

File: azure/services/publicips/publicips.go [hunk 5]
```diff
@@ -149,3 +149,29 @@ func (s *Service) isIPManaged(ctx context.Context, ipName string) (bool, error)
 	tags := converters.MapToTags(ip.Tags)
 	return tags.HasOwned(s.Scope.ClusterName()), nil
 }
+
+func getPublicIPAddressPropertiesFormat(addressVersion network.IPVersion, dnsSettings *network.PublicIPAddressDNSSettings, outboundIP, outboundIPPrefix net.IP) *network.PublicIPAddressPropertiesFormat {
+	if outboundIP != nil {
```
Comment (shysank) [inline azure/services/publicips/publicips.go:154]: can we set the common fields outside of the conditions and append outbound ip or prefix based on the inputs?

Comment (shysank) [reply azure/services/publicips/publicips.go:154]: I'd also suggest moving the parsing logic inside the helper and parse lazily only if `ip.OutboundIP != nil`

Comment (shysank) [reply azure/services/publicips/publicips.go:154]: something like 
```
pipAddressProperties := &network.PublicIPAddressPropertiesFormat{
			PublicIPAddressVersion:   addressVersion,
			PublicIPAllocationMethod: network.IPAllocationMethodStatic,
			DNSSettings:              dnsSettings,
		}

if ip.outboundip == nil {
return pipAddressProperties
}

// append outbound ip/prefix here

```

File: api/v1alpha4/azurecluster_conversion.go [hunk 6]
```diff
@@ -40,6 +40,38 @@ func (src *AzureCluster) ConvertTo(dstRaw conversion.Hub) error { // nolint
 	// Restore list of virtual network peerings
 	dst.Spec.NetworkSpec.Vnet.Peerings = restored.Spec.NetworkSpec.Vnet.Peerings
 
+	if restored.Spec.NetworkSpec.NodeOutboundLB != nil {
+		for i, _ := range restored.Spec.NetworkSpec.NodeOutboundLB.FrontendIPs {
+			dst.Spec.NetworkSpec.NodeOutboundLB.FrontendIPs[i].PublicIP = restored.Spec.NetworkSpec.NodeOutboundLB.FrontendIPs[i].PublicIP
```
Comment (shysank) [inline api/v1alpha4/azurecluster_conversion.go:45]: shouldn't we only set publicIP.outboundIP?

File: api/v1beta1/azurecluster_validation.go [hunk 7]
```diff
@@ -437,6 +437,16 @@ func validateNodeOutboundLB(lb *LoadBalancerSpec, old *LoadBalancerSpec, apiserv
 			fmt.Sprintf("Node outbound idle timeout should be between %d and %d minutes", MinLBIdleTimeoutInMinutes, MaxLoadBalancerOutboundIPs)))
 	}
 
+	for index, frontendIP := range lb.FrontendIPs {
```
Comment (shysank) [inline api/v1beta1/azurecluster_validation.go:440]: Although we intend to allow outboundIP only for nodeoutboundLB, there is nothing stopping the user from setting it in other places where publicIP is used (eg. AzureBastion). Should we add validations there to prevent the user from setting it?

File: azure/services/publicips/publicips.go [hunk 8]
```diff
@@ -75,6 +76,9 @@ func (s *Service) Reconcile(ctx context.Context) error {
 			}
 		}
 
+		outboundIP := net.ParseIP(ip.OutboundIP)
```
Comment (shysank) [inline azure/services/publicips/publicips.go:79]: We should add mapping from api to scope for the new field `outboundIP` => https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/45519b18be2cb7a26034419df99d46c87b410285/azure/scope/cluster.go#L125

File: api/v1beta1/azurecluster_validation_test.go (hunk 2 above)
Comment (Ankitasw) [reply api/v1beta1/azurecluster_validation_test.go:1028]: no will remove it, was added in previous changes

File: api/v1beta1/azurecluster_validation_test.go (hunk 3 above)
Comment (Ankitasw) [reply api/v1beta1/azurecluster_validation_test.go:1031]: same as above

File: api/v1beta1/azurecluster_validation.go (hunk 4 above)
Comment (Ankitasw) [reply api/v1beta1/azurecluster_validation.go:518]: sounds fair to me

File: api/v1alpha4/azurecluster_conversion.go (hunk 6 above)
Comment (Ankitasw) [reply api/v1alpha4/azurecluster_conversion.go:45]: missed it, will add only outboundIP here

File: api/v1beta1/azurecluster_validation.go (hunk 7 above)
Comment (Ankitasw) [reply api/v1beta1/azurecluster_validation.go:440]: I notices PublicIP is used only in AzureBastion other than loadbalancer spec, do you know any other places where it is used or I have missed?

File: azure/services/publicips/publicips.go (hunk 8 above)
Comment (Ankitasw) [reply azure/services/publicips/publicips.go:79]: something like [this](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/57ca45a908419f8ae0ed2a78e8bacc53db25a41f/azure/scope/cluster.go#L734) ?

File: azure/services/publicips/publicips.go (hunk 5 above)
Comment (Ankitasw) [reply azure/services/publicips/publicips.go:160]: fixed now

File: azure/services/publicips/publicips.go (hunk 8 above)
Comment (shysank) [reply azure/services/publicips/publicips.go:79]: yeah something like that. To verify the expected behavior, I'd suggest creating a cluster using tilt and see if the appropriate resources are created.

Comment (Ankitasw) [reply azure/services/publicips/publicips.go:79]: ok but how to use existing IP, I mean which IPs can I use for testing?
